src/pudl/analysis/allocate_net_gen.py

```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+
+    logger.info(
+        "Ratio calc types: \n"
+        f"   All gens w/in generation table:  {len(all_gen)}#, {all_gen.capacity_mw.sum():.2} MW\n"
+        f"   Some gens w/in generation table: {len(some_gen)}#, {some_gen.capacity_mw.sum():.2} MW\n"
+        f"   No gens w/in generation table:   {len(bf_only)}#, {bf_only.capacity_mw.sum():.2} MW"
+    )
+    if len(gen_pm_fuel) != len(all_gen) + len(some_gen) + len(bf_only):
+        raise AssertionError(
+            "Error in splitting the gens between records showing up fully, "
+            "partially, or not at all in the generation table."
+        )
+
+    # In the case where we have all of the fuel from the bf
+    # table, we still allocate, because the fuel reported in these two
+    # tables don't always match perfectly
+    all_gen = all_gen.assign(
+        frac_fuel=lambda x: x.fuel_consumed_mmbtu_bf_tbl
+        / x.fuel_consumed_mmbtu_bf_tbl_pm_fuel,
+        frac=lambda x: x.frac_fuel,
+    )
+    # _ = _test_frac(all_gen)
+
+    some_gen = some_gen.assign(
+        # fraction of the fuel consumption that should go to the generators that
+        # report in the boiler fuel table
+        frac_from_bf_tbl=lambda x: x.fuel_consumed_mmbtu_bf_tbl_pm_fuel
+        / x.fuel_consumed_mmbtu_gf_tbl,
+        # for records within these mix groups that do have fuel consumption in the
+        # bf table..
+        frac_fuel=lambda x: x.fuel_consumed_mmbtu_bf_tbl
+        / x.fuel_consumed_mmbtu_bf_tbl_pm_fuel,  # generator based fuel from bf table
+        frac_bf=lambda x: x.frac_fuel * x.frac_from_bf_tbl,
+        # fraction of fuel that does not show up in the bf table
+        # set minimum fraction to zero so we don't get negative fuel
+        frac_missing_from_bf_tbl=lambda x: np.where(
+            (x.frac_from_bf_tbl < 1), (1 - x.frac_from_bf_tbl), 0
+        ),
+        capacity_mw_missing_from_bf_tbl=lambda x: np.where(
+            x.in_bf_tbl, 0, x.capacity_mw
+        ),
+        frac_cap=lambda x: x.frac_missing_from_bf_tbl
+        * (x.capacity_mw_missing_from_bf_tbl / x.capacity_mw_fuel_in_bf_tbl_group),
+        # the real deal
+        # this could aslo be `x.frac_bf + x.frac_cap` because the frac_bf
+        # should be 0 for any generator that does not have fuel in the bf_tbl
+        # and frac_cap should be 0 for any generator that has fuel in the
+        # bf_tbl.
+        frac=lambda x: np.where(x.in_bf_tbl, x.frac_bf, x.frac_cap),
+    )
+    # _ = _test_frac(some_gen)
+
+    # Calculate what fraction of the total capacity is associated with each of
+    # the generators in the grouping.
+    bf_only = bf_only.assign(
+        frac_cap=lambda x: x.capacity_mw / x.capacity_mw_unit_fuel,
+        frac=lambda x: x.frac_cap,
+    )
+    # _ = _test_frac(bf_only)
+
+    # squish all of these methods back together.
+    fuel_alloc = pd.concat([all_gen, some_gen, bf_only])
+    # null out the inf's
+    fuel_alloc.loc[abs(fuel_alloc.frac) == np.inf] = np.NaN
```

Comment:
re: my 0's/nulls comment before I think I like this solution in general, but if we are going to use the small numbers method to deal with zeros/nulls do we need this at all?